### PR TITLE
SAK-39981: The final course grade override fails

### PR DESCRIPTION
**Pull request by raulsv**

### CODE REVIEW

File: edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradeMappingDefinition.java [hunk 1]
```diff
@@ -102,9 +103,10 @@ public String toString() {
 
 		// we only ever order by bottom percents now
 		final DoubleComparator doubleComparator = new DoubleComparator(gradeMap);
-		final Map<String, Double> rval = new TreeMap<>(doubleComparator);
-		rval.putAll(gradeMap);
-
+		final Map<String, Double> tMap = new TreeMap<>(doubleComparator);
+		tMap.putAll(gradeMap);
+		final Map<String, Double> rval = new LinkedHashMap<>(tMap);
```
Comment (steveswinsburg) [inline edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradeMappingDefinition.java]: Can you just double check where this method is used as I commented in the javadoc for it (now out of date) that it was specifically a treemap and to ensure that was considered. There might be additional unnecessary casts down the track.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java [hunk 2]
```diff
@@ -962,7 +962,7 @@ public void putCourseGradesInMatrix(Map<String, GbStudentGradeInfo> matrix, List
 
 		// Setup the course grade formatter
 		// TODO we want the override except in certain cases. Can we hard code this?
-		final CourseGradeFormatter courseGradeFormatter = new CourseGradeFormatter(gradebook, role, isCourseGradeVisible, settings.getShowPoints(), false);
+		final CourseGradeFormatter courseGradeFormatter = new CourseGradeFormatter(gradebook, role, isCourseGradeVisible, settings.getShowPoints(), true);
```
Comment (steveswinsburg) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: why was this ever false? It means to use the course grade override when showing the data.

Comment (ern) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: Looks like a regression from SAK-33849 #5245

Comment (steveswinsburg) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: @bjones86 was there a reason for the change from true to false in the above PR or is this a regression?

Comment (bjones86) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java]: Looks like this change was accidental.

File: edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradeMappingDefinition.java (hunk 1 above)
Comment (raulsv) [reply edu-services/gradebook-service/api/src/java/org/sakaiproject/service/gradebook/shared/GradeMappingDefinition.java]: Hi Steve

I've searched for a "cast" on all methods where the "sortGradeMapping" method is called and I've found no "cast". The "sortGradeMapping" method return is only used as a Map and no implementation is used. The "sortGradeMapping" is called in these methods:

- org.sakaiproject.component.gradebook.GradebookServiceHibernateImpl.getCourseGradeForStudents(String,
List<String>, Map<String, Double>)

- org.sakaiproject.gradebookng.tool.component.GbCourseGradeChart.getData(Map<String, Double>)

- org.sakaiproject.component.gradebook.GradebookServiceHibernateImpl.getGradebookGradeMappings (Set<GradeMapping>)

- org.sakaiproject.component.gradebook.GradebookServiceHibernateImpl.getGradebookInformation(String)

- org.sakaiproject.gradebookng.tool.panels.SettingsGradingSchemaPanel.getGradingSchemaEntries()

- org.sakaiproject.gradebookng.tool.panels.SettingsGradingSchemaPanel.refreshCourseGradeChart
(AjaxRequestTarget)

**Comment by raulsv** [comment]
@steveswinsburg Could you take a look at the updated comment? Thanks

**Comment by danielmerino** [comment]
Hi @steveswinsburg  S2U team has determined that this change is low risk. If you have no objection, we will do a merge in our next meeting (next week).

**Comment by steveswinsburg** [comment]
Sorry for the delay. Will get this in and re-verify the grading schema sorting. Thanks.

**Pull request merged**